src/profiler/base/profiler.cc

```diff
@@ -49,6 +49,19 @@ std::string Profiler::GetProfile() {
     profile_stat->EmitEvents(&ss);
     ++stat_count;
   }
+  if (stat_count == 0 && !helpers_.empty()) {
+    CollectStat();
+    while (profile_stats_.opr_exec_stats_->try_dequeue(stat)) {
+      CHECK_NOTNULL(stat);
+      std::unique_ptr<ProfileStat> profile_stat(stat);  // manage lifecycle
+      CHECK_NE(profile_stat->categories_.c_str()[0], '\0') << "Category must be set";
+      if (stat_count) {
+        ss << ",\n";
+      }
+      profile_stat->EmitEvents(&ss);
+      ++stat_count;
```

Comment:
I feel the code is little bit redundant.  You can have a check before line 42 to see if there is elements in the queue, probably use `size_approx()`. If no element, and `helpers_` is not empty, then perform `CollectStat()`.    The code could be concise.

Reply:
Previously not sure whether `size_approx()` would always return a positive number when the queue is not empty. 

Read the source code in dmlc again. Since we use `std::lock_guard` and all memory operations are settled before accessing the queue, I think the estimation of queue size using `size_approx()` would be accurate.

Reply:
I am not 100% sure about that. If it does not work correctly, you can check whether the first `stat` is null. Anyway, I think we can verify that with more tests.